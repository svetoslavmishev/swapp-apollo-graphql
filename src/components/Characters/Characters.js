import React, { useContext } from 'react';
import { useQuery } from '@apollo/react-hooks';
import { useHistory } from 'react-router-dom';
import { Button } from '@material-ui/core';

import { Header, CharCard } from '../index';
import Loading from '../shared/Loading/Loading';
import { MORE_CHARACTERS } from '../../queries/queries';
import { ThemeContext } from '../../themeContext';
import styles from './CharactersStyles';

const Characters = () => {
  const { currentTheme } = useContext(ThemeContext);
  const classes = styles({ currentTheme });
  const { data, loading, error, fetchMore } = useQuery(MORE_CHARACTERS);
  const history = useHistory();

  const loadMore = () => {
    fetchMore({
      variables: {
        cursor: data.allPeople.pageInfo.endCursor
      },
      updateQuery: (previousResult, { fetchMoreResult: { allPeople } }) => {
        const newEdges = allPeople.edges;
        const pageInfo = allPeople.pageInfo;

        return newEdges.length
          ? {
              allPeople: {
                __typename: previousResult.allPeople.__typename,
                edges: [...previousResult.allPeople.edges, ...newEdges],
                pageInfo
              }
            }
          : previousResult;
      }
    });
  };

  if (loading) return <Loading />;
  if (error) return null;

  return (
    <>
      <Header />
      <div className={classes.root}>
        <div className="flex flex-wrap justify-around p-4">
          {data &&
            data.allPeople.edges.map(char => {
              return (
                <CharCard
                  key={char.node.id}
                  chars={char.node}
                  onClick={() => history.push(`/characters/${char.node.id}`)}
                />
              );
            })}
        </div>
        <div className="flex justify-center">
          <Button
            className={classes.button}
            variant="contained"
            onClick={loadMore}
          >
            Load more
          </Button>
        </div>
      </div>
    </>
  );
};

export default Characters;
