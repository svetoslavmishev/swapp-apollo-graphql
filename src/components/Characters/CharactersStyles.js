import { makeStyles } from '@material-ui/styles';

const styles = makeStyles(theme => ({
  root: {
    minHeight: 'calc(100vh - 64px)',
    background: ({ currentTheme }) =>
      theme.palette[currentTheme].backgroundColor
  },
  button: {
    fontFamily: 'StarWars',
    background: ({ currentTheme }) =>
      theme.palette[currentTheme].SolidButtons.backgroundColor,
    color: ({ currentTheme }) =>
      theme.palette[currentTheme].SolidButtons.fontColor,
    margin: 40,
    '&:hover': {
      color: '#000000',
      background: theme.palette.commonColors.yellow
    }
  }
}));

export default styles;
