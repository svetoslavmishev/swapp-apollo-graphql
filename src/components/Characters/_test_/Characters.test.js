import React from 'react';
import { shallow } from 'enzyme';
import Characters from '../Characters';
import { ApolloProvider } from '@apollo/react-hooks';

describe('<Characters />', () => {
  it('renders <Characters /> component without error', () => {
    const wrapper = shallow(
      <ApolloProvider>
        <Characters />
      </ApolloProvider>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
