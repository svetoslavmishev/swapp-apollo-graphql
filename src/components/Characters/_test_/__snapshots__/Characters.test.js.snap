// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Characters /> renders <Characters /> without error 1`] = `ShallowWrapper {}`;
